// glitched-writer/lib/cjs/modules/state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
class State {
    constructor(writer) {
        this.nGhosts = 0;
        /**
         * Numerical data about progress of writing
         */
        this.progress = {
            percent: 0,
            done: 0,
            todo: 0,
            increase() {
                this.done++;
                this.percent = this.done / this.todo;
            },
            reset(todo) {
                this.percent = 0;
                this.done = 0;
                this.todo = todo;
            },
            finish() {
                this.done = this.todo;
                this.percent = 1;
            },
        };
        this.isTyping = false;
        this.isPaused = false;
        this.finished = true;
        this.erasing = false;
        this.writer = writer;
        this.maxGhosts = this.writer.options.genMaxGhosts;
    }
    get ghostsInLimit() {
        return this.nGhosts < this.maxGhosts;
    }
    play() {
        this.isTyping = true;
        this.isPaused = false;
        this.finished = false;
        this.toggleClass(true);
        this.maxGhosts = this.writer.options.genMaxGhosts;
    }
    pause() {
        this.isTyping = false;
        this.isPaused = true;
        this.toggleClass(false);
    }
    finish() {
        this.progress.finish();
        this.isTyping = false;
        this.finished = true;
        this.toggleClass(false);
    }
    toggleClass(enable) {
        const el = this.writer.htmlElement, className = 'gw-writing';
        enable ? utils_1.animateWithClass(el, className) : el.classList.remove(className);
    }
}
exports.default = State;
